File: agents/AgentServer/AgentServer.js
const express = require('express');
const http = require('http');
const WebSocket = require('ws');

class AgentServer {
  constructor(port = 4242) {
    this.port = port;
    this.app = express();
    this.server = http.createServer(this.app);
    this.wss = new WebSocket.Server({ server: this.server });
  }

  start() {
    this.initializeHTTP();
    this.initializeWebSocket();
    this.server.listen(this.port, () => {
      console.log(`AgentServer listening on port ${this.port}`);
    });
  }

  initializeHTTP() {
    this.app.use(express.json());
    this.app.post('/message', (req, res) => {
      const message = req.body;
      const transport = new HTTPTransport(res);
      this.handleIncomingMessage(message, transport);
    });
  }

  initializeWebSocket() {
    this.wss.on('connection', (ws) => {
      const transport = new WebSocketTransport(ws);
      ws.on('message', (message) => {
        this.handleIncomingMessage(JSON.parse(message), transport);
      });
    });
  }

  handleIncomingMessage(message, transport) {
    if ('task' in message) {
      this.handleTaskRequest(message, transport);
    } else if ('current_agent_ID' in message) {
      this.handleHandoffNotification(message, transport);
    } else if ('answer' in message) {
      this.handleClarificationResponse(message, transport);
    } else if ('status' in message) {
      this.handleCallbackNotification(message, transport);
    } else {
      const errorMessage = {
        error_message: 'Unknown message type',
        requesting_agent_ID: message.requesting_agent_ID || 'unknown',
      };
      this.sendResponse(errorMessage, transport);
    }
  }

  sendResponse(response, transport) {
    transport.sendResponse(response);
  }

  handleTaskRequest(taskRequest, transport) {
    // To be overridden by child classes
    console.log('Received task request:', taskRequest);
  }

  handleHandoffNotification(handoffNotification, transport) {
    // To be overridden by child classes
    console.log('Received handoff notification:', handoffNotification);
  }

  handleCallbackNotification(callbackNotification, transport) {
    // To be overridden by child classes
    console.log('Received callback notification:', callbackNotification);
  }

  handleClarificationResponse(clarificationResponse, transport) {
    // To be overridden by child classes
    console.log('Received clarification response:', clarificationResponse);
  }
}

class HTTPTransport {
  constructor(res) {
    this.res = res;
  }

  sendResponse(response) {
    this.res.json(response);
  }
}

class WebSocketTransport {
  constructor(ws) {
    this.ws = ws;
  }

  sendResponse(response) {
    this.ws.send(JSON.stringify(response));
  }
}

// Type definitions
/**
 * @typedef {Object} TaskRequest
 * @property {string} task_ID
 * @property {string} task
 * @property {string} requesting_agent_ID
 * @property {string} [callback_URL]
 */

/**
 * @typedef {Object} TaskResponse
 * @property {string} task_ID
 * @property {string} result
 * @property {('success'|'failed'|'in-progress')} status
 * @property {string} replying_agent_ID
 */

/**
 * @typedef {Object} HandoffNotification
 * @property {string} task_ID
 * @property {string} current_agent_ID
 * @property {string} new_agent_ID
 * @property {string} reason_for_handoff
 */

/**
 * @typedef {Object} ClarificationRequest
 * @property {string} task_ID
 * @property {string} question
 * @property {string} requesting_agent_id
 */

/**
 * @typedef {Object} ClarificationResponse
 * @property {string} task_ID
 * @property {string} answer
 * @property {string} replying_agent_id
 */

/**
 * @typedef {Object} CallbackNotification
 * @property {string} task_ID
 * @property {string} result
 * @property {('complete'|'failed'|'in-progress')} status
 * @property {string} requesting_agent_ID
 * @property {string} replying_agent_ID
 */

/**
 * @typedef {Object} ErrorMessage
 * @property {string} [task_ID]
 * @property {string} error_message
 * @property {string} requesting_agent_ID
 */

/**
 * @typedef {Object} TransportInterface
 * @property {function(any): void} sendResponse
 */

module.exports = {
  AgentServer,
  // Export type definitions (these are just for documentation purposes in JavaScript)
  TaskRequest: null,
  TaskResponse: null,
  HandoffNotification: null,
  ClarificationRequest: null,
  ClarificationResponse: null,
  CallbackNotification: null,
  ErrorMessage: null,
  TransportInterface: null,
};

File: agents/AgentServer/AgentServer.py
from pydantic import BaseModel, Field
from typing import Optional, Union
from enum import Enum
import asyncio
import websockets
from aiohttp import web
import json

class TaskStatus(str, Enum):
    SUCCESS = "success"
    FAILED = "failed"
    IN_PROGRESS = "in-progress"

class TaskRequest(BaseModel):
    task_ID: str
    task: str
    requesting_agent_ID: str
    callback_URL: Optional[str] = None

class TaskResponse(BaseModel):
    task_ID: str
    result: str
    status: TaskStatus
    replying_agent_ID: str

class HandoffNotification(BaseModel):
    task_ID: str
    current_agent_ID: str
    new_agent_ID: str
    reason_for_handoff: str

class ClarificationRequest(BaseModel):
    task_ID: str
    question: str
    requesting_agent_id: str

class ClarificationResponse(BaseModel):
    task_ID: str
    answer: str
    replying_agent_id: str

class CallbackNotification(BaseModel):
    task_ID: str
    result: str
    status: TaskStatus
    requesting_agent_ID: str
    replying_agent_ID: str

class ErrorMessage(BaseModel):
    task_ID: Optional[str] = None
    error_message: str
    requesting_agent_ID: str

class TransportInterface:
    async def sendResponse(self, response: Union[TaskResponse, ClarificationResponse, CallbackNotification, ErrorMessage]):
        raise NotImplementedError()

class WebSocketTransport(TransportInterface):
    def __init__(self, websocket):
        self.websocket = websocket

    async def sendResponse(self, response: Union[TaskResponse, ClarificationResponse, CallbackNotification, ErrorMessage]):
        await self.websocket.send(response.json())

class HTTPTransport(TransportInterface):
    def __init__(self, response):
        self.response = response

    async def sendResponse(self, response: Union[TaskResponse, ClarificationResponse, CallbackNotification, ErrorMessage]):
        self.response.text = response.json()
        self.response.content_type = 'application/json'

class AgentServer:
    def __init__(self, port: int = 4242):
        self.port = port
        self.app = web.Application()
        self.app.router.add_post('/task', self.http_task_handler)

    async def start(self):
        self.server = await asyncio.start_server(self.websocket_handler, '0.0.0.0', self.port)
        runner = web.AppRunner(self.app)
        await runner.setup()
        self.site = web.TCPSite(runner, '0.0.0.0', self.port)
        await self.site.start()
        print(f"Server started on port {self.port}")

    async def websocket_handler(self, websocket, path):
        transport = WebSocketTransport(websocket)
        async for message in websocket:
            await self.handleIncomingMessage(json.loads(message), transport)

    async def http_task_handler(self, request):
        data = await request.json()
        response = web.Response()
        transport = HTTPTransport(response)
        await self.handleIncomingMessage(data, transport)
        return response

    async def handleIncomingMessage(self, message: dict, transport: TransportInterface):
        try:
            if 'task' in message:
                await self.handleTaskRequest(TaskRequest(**message), transport)
            elif 'reason_for_handoff' in message:
                await self.handleHandoffNotification(HandoffNotification(**message), transport)
            elif 'question' in message:
                await self.handleClarificationResponse(ClarificationResponse(**message), transport)
            elif 'status' in message:
                await self.handleCallbackNotification(CallbackNotification(**message), transport)
            else:
                raise ValueError("Unknown message type")
        except Exception as e:
            error = ErrorMessage(error_message=str(e), requesting_agent_ID="unknown")
            await self.sendResponse(error, transport)

    async def sendResponse(self, response: Union[TaskResponse, ClarificationResponse, CallbackNotification, ErrorMessage], transport: TransportInterface):
        await transport.sendResponse(response)

    async def handleTaskRequest(self, task_request: TaskRequest, transport: TransportInterface):
        # This method should be overridden in a subclass to implement task processing
        pass

    async def handleHandoffNotification(self, handoff: HandoffNotification, transport: TransportInterface):
        # This method should be overridden in a subclass to implement handoff processing
        pass

    async def handleCallbackNotification(self, callback: CallbackNotification, transport: TransportInterface):
        # This method should be overridden in a subclass to implement callback processing
        pass

    async def handleClarificationResponse(self, clarification: ClarificationResponse, transport: TransportInterface):
        # This method should be overridden in a subclass to implement clarification processing
        pass

if __name__ == "__main__":
    server = AgentServer()
    asyncio.run(server.start())
    try:
        asyncio.get_event_loop().run_forever()
    except KeyboardInterrupt:
        print("Server stopped")